Refactor to work with React 0.14 (and older)

JakeSidSmith opened the pull request:
- Modify elements that have `onClick` props with additional event props
- Decrease chance of memory leaks
- Fixes #13 (removing, hiding and moving elements still triggering click events)
- Click events only triggered if touch released on same element
- Allow standard onClick events to happen (allowing checkboxes, labels, inputs, focusing, etc to function)
- Handle special cases for inputs etc be focussed / accessed quickly
- Remove need for timeouts
- Handle `React.DOM.x` element creation
- Support 0.14+ (fixes #14)
- Support older versions of React (theoretically - not tested)

**New features (mentioned in / fixes #4)**
- Added `event.button = 0`
- Changed `event.type` to `click`
- Added `event.fastclick = true`
- Copy touch positions to event

**Need testers!**
Update your package.json

```
"react-fastclick": "git://github.com/jakesidsmith/react-fastclick.git#refactor-for-0.14"
```

Include react-fastclick

```
require('react-fastclick');
// or if you're using ES6
import 'react-fastclick';
```

Please post what version of react you've tested with, what devices / browsers you've tested on, which elements / components (if any) you have had problems with, and what libraries may be affecting this.

JakeSidSmith committed:
Basic fastclick wrapper (draft)

JakeSidSmith committed:
Fix linting

JakeSidSmith committed:
Remove wrapper class

JakeSidSmith committed:
Apply to React.DOM methods

JakeSidSmith committed:
Remove console logs

JakeSidSmith committed:
Remove cancel handler

JakeSidSmith committed:
Always reset moved variable

JakeSidSmith committed:
Moved can still be set if touches invalid

JakeSidSmith committed:
No need to check for moved again & always call touchEnd callback

JakeSidSmith committed:
Reset moved onClick

JakeSidSmith committed:
Update comment

JakeSidSmith committed:
Update comment

JakeSidSmith committed:
Freeze props if possible

JakeSidSmith committed:
Freeze if freeze is a function

JakeSidSmith committed:
Handle input & label special behaviour

JakeSidSmith committed:
Focus text areas and selects

JakeSidSmith committed:
Do not prevent default on selects

JakeSidSmith committed:
Prevent default only for inputs of type checkbox or radio

JakeSidSmith committed:
Only apply to elements with onClick handlers or of special types

JakeSidSmith committed:
Rename focus and check function

slybridges commented:
Hi @JakeSidSmith,

I stumbled upon your repo as I was looking for an alternative to `fastclick`. Fastclick doesn't work properly natively with `react-select` on mobile.

Your PR makes it work correctly, so yay ! :)

However there seem to be a conflict with the `Link` component of `react-router`: when I navigate to the app and click on `Link` components, the app performs a full refresh of the page.

If I remove `react-fastclick` Links work properly again (but indeed, the whole app feels sluggish on mobile).

Some info:
- Tested on Safari Mobile, on IOS 9.2
- `react v0.14.6`
- `react-router: 2.0.0-rc4` using `browserHistory`

JakeSidSmith commented:
Thanks for testing this out @slybridges. I thought I had fixed the issues with react-router, but I guess not. Will take a look later today. :)

JakeSidSmith commented:
@slybridges Have just tried this branch out with 2 different projects that use `react-router`, and I don't have this problem. Though they are running `react-router` versions `0.13` and `1.0.0`. Could you try clearing `react-fastclick` from your `node_modules` and rebuild. If you still experience that issue, I'll create a new project with `react-router` `2.0.0` and try to debug. :)

slybridges commented:
Hi @JakeSidSmith unfortunately clearing the module from node-modules didn't help.

I was able to reproduce the issue using the`redux-easy-boilerplate` repo https://github.com/anorudes/redux-easy-boilerplate).

You'll find a version ready to be tested along with my observations there: https://github.com/slybridges/redux-easy-boilerplate/tree/master

As you'll notice, problem seems to be with using `react-fastclick` along with `createHistory` (the one using HTML5 API). Using `createHashHistory` (the one using legacy '#' keyed URLs), no problem arises. And using `createHistory` without `react-fastclick` works too.

Hope this can be of some help for you!

JakeSidSmith committed:
Fake click event (fixes react-router issue)

JakeSidSmith commented:
Hey, @slybridges think it's fixed now.

The link component was checking for the button that was clicked (to make sure it wasn't a right click), so I'm just adding this to the event - `event.button = 0;`.

I've also modified the `event.type` to be `click` to avoid other libraries that may check for this, and added a `fastclick` key so users scan still tell the difference as mentioned in #4.

Will do some additional work to push in clientX, pageX, etc so that people that are expecting an actual click event can still access all the usual values, and then this branch should be good to go. :)

slybridges commented:
Hi @JakeSidSmith, it seems to be working alright for me now thanks ! Although I'm yet to deploy to production environnement with it and asses the benefits :)

JakeSidSmith committed:
Copy position keys from touch

JakeSidSmith committed:
Include radius in position over element check

JakeSidSmith committed:
Missing semicolon

JakeSidSmith committed:
Increment major version number

JakeSidSmith committed:
Update description

JakeSidSmith committed:
Update support info

JakeSidSmith committed:
Update examples

JakeSidSmith committed:
Update notes about events

JakeSidSmith committed:
Usage info update

JakeSidSmith committed:
Warn about using older versions

JakeSidSmith committed:
Update support text

JakeSidSmith committed:
Update linting & code styling dependencies

manosim commented:
@JakeSidSmith Looks awesome :sparkles:

glenjamin commented:
Awesome 🤑
